_source/_posts/2021-12-03-angular-dynamic-components.md

```diff
@@ -0,0 +1,773 @@
+
+Open `src/app/protected/department/dynamic.component.ts` and add a property named `data` of type `MessageData` to it. The interface now looks like the following.
+
+```ts
+export interface DynamicComponent {  
+  data: MessageData;  
+}
+```
+
+Because we have separate components specifically designed to handle each message type, we keep the logic very straightforward. We'll apply template updates to each component and implement all the interface members by adding an input property to accept the data.
+
+Start by opening `src/app/protected/department/clawesome.component.ts`. This component's data has an URL to an image and string content.  Update the component to the following.
+
+{% raw %}
+```ts
+@Component({  
+  selector: 'app-clawesome',  
+  template:  `  
+    <mat-card class="card">
+      <img mat-card-image src="{{data.url}}" alt="Photo of a clawesome creature" >  
+      <mat-card-content>
+        <p>{{data.content}}</p>  
+      </mat-card-content>
+    </mat-card>
+  `,  
+  styles: [` .card { max-width: 300px; } `]  
+})  
+export class ClawesomeComponent implements DynamicComponent {  
+ @Input() data!: MessageData;  
+}
+```
+{% endraw %}
+
+Each component is a Material card control with an image to display. Next open `src/app/protected/department/pawesome.component.ts`. In addition to the URL, content contains the properties `name` and `about`. Update the component to the following.
+
+{% raw %}
+```ts
+@Component({  
+  selector: 'app-pawesome',  
+  template: `  
+    <mat-card class="card">
+      <mat-card-header>
+        <mat-card-title>{{data.content.name}}</mat-card-title>  
+        <mat-card-subtitle>Good doggo</mat-card-subtitle>
+      </mat-card-header>
+      <img mat-card-image src="{{data.url}}" alt="Photo of a pawesome creature" >  
+      <mat-card-content>
+        <p> {{data.content.about}} </p>
+      </mat-card-content>
+    </mat-card>
+  `,  
+  styles: [` .card { max-width: 300px; } `]  
+})  
+export class PawesomeComponent implements DynamicComponent {  
+ @Input() data!: MessageData;  
+}
+```
+{% endraw %}
+
+Finally, open `src/app/protected/department/smiley.component.ts`. The only data in this message type is the URL. Update the component to the following.
+
+{% raw %}
+```ts
+@Component({  
+  selector: 'app-smiley',  
+  template: `  
+    <mat-card class="card">
+      <img mat-card-image src="{{data.url}}" alt="Photo of a smiley creature" >  
+      <mat-card-content>
+        <p>SMILE!</p>
+      </mat-card-content>
+    </mat-card> `,  
+  styles: [` .card { max-width: 300px; } `]
+})  
+export class SmileyComponent implements DynamicComponent {  
+ @Input() public data!: MessageData;  
+}
+```
+{% endraw %}
+
+The rotating components look a little goofy now and you'll see a lot of errors in the console, but we'll get it fixed up here soon.
```

Comment:
Is this line gramattically correct?

Reply:
Maybe a little folksy, but I think correct (also ran through Grammarly). I edited it to '... but we'll get it fixed up soon.' so that it's less folksy. Does that help?